### PR TITLE
Match up MP for year of stats correctly from postcode

### LINKED ISSUE #247: Responsiveness table postcode look-up not working where MPs have changed
https://www.writetothem.com/stats/2013/mps

eg  CF14 2FN results in an error message when it should show the ex-MP Jonathan Evans.

### PR DESCRIPTION
**Pull request by struan**
Previously we were looking for the stats for the current MP who may
not have been MP in previous years. This was returning an error. We
now fetch all the reps for an area and then loop over them till we
find one who is both in that years results and has a matching
constituency name.

Fixes #247

### CODE REVIEW

File: web/stats.php [hunk 1]
```diff
@@ -32,6 +32,22 @@
     $previous_year = 'FYMP';
 
 $years = array('2005', '2006', '2007', '2008', '2013', '2014');
+
+# this is so we get the mapit IDs and cons names for the year and
+# hence can get the MP for the year from DaDem and not the current
+# on
+$mapit_year_args = array(
```
Comment (dracos) [inline web/stats.php]: Better to do this with range checks? Then you won't have to add anything for 2015.

File: web/stats.php [hunk 2]
```diff
@@ -53,16 +69,18 @@
 require_once "../phplib/questionnaire_report_${year}_WMC.php";
 
 $rep_info = array();
-$voting_areas = mapit_call('postcode', $postcode, array(), array(
+$area_representatives = array();
+$area_name = '';
+$voting_areas = mapit_call('postcode', $postcode, $mapit_args, array(
```
Comment (dracos) [inline web/stats.php]: I realise this doesn't change behaviour, but it looks like this call happens even if the postcode is blank, which seems a bit superfluous :)

File: web/stats.php (hunk 1 above)
Comment (struan) [inline web/stats.php]: oh, possibly yes.

File: web/stats.php [hunk 3]
```diff
@@ -53,16 +69,18 @@
 require_once "../phplib/questionnaire_report_${year}_WMC.php";
 
 $rep_info = array();
```
Comment (dracos) [inline web/stats.php]: rep_info could be initialized down in the type == mps bit now.

File: web/stats.php [hunk 4]
```diff
@@ -80,6 +98,27 @@
...
+    # 2007 would get you Boris Johnson's stats for Henley in 2007. This only works if
+    # the cons name matches with MaPit which is should :/
+    if ($area_representatives) {
+        foreach ( $area_representatives as $rep ) {
+            $rep_info = dadem_get_representative_info($rep);
+            dadem_check_error($rep_info);
+            $key = $rep_info['parlparse_person_id'];
+            if ( array_key_exists($key, $GLOBALS["questionnaire_report_${year}_WMC"]) &&
+                 $area_name == $GLOBALS["questionnaire_report_${year}_WMC"][$key]["area"] ) {
+                $rep_info['postcode'] = $postcode;
```
Comment (dracos) [inline web/stats.php]: Given #314 changes, this line won't need to be set any more either.

File: web/stats.php [hunk 5]
```diff
@@ -80,6 +98,27 @@
...
+    # the current MP may not be the MP from the year of the stats so loop back
+    # through the reps till we get one who is in the stats. We need to check the
+    # area is the same on the small chance that e.g when asking for the 2007 stats
+    # the MP elected in 2015 was MP for a different seat and so you get their stats but
+    # for the wrong seat - e.g asking for stats for Uxbridge and South Ruislip for
+    # 2007 would get you Boris Johnson's stats for Henley in 2007. This only works if
+    # the cons name matches with MaPit which is should :/
+    if ($area_representatives) {
+        foreach ( $area_representatives as $rep ) {
+            $rep_info = dadem_get_representative_info($rep);
```
Comment (dracos) [inline web/stats.php]: Perhaps more straightforward to set something else here, and then if it matches set rep_info and break? Then you don't need to empty it on each iteration.

**Comment by dracos** [comment]
Few small things, otherwise fine.

**Pull request merged**